pkg/block/fetcher_test.go

```diff
@@ -393,3 +401,271 @@ func TestTimePartitionMetaFilter_Filter(t *testing.T) {
+	}{
+		{
+			name: "3 non compacted metas in bucket",
+			input: map[ulid.ULID]*metadata.Meta{
+				ULID(1): {
+					BlockMeta: tsdb.BlockMeta{
+						ULID: ULID(1),
+						Compaction: tsdb.BlockMetaCompaction{
+							Sources: []ulid.ULID{ULID(1)},
+						},
+					},
+				},
+				ULID(2): {
+					BlockMeta: tsdb.BlockMeta{
+						ULID: ULID(2),
+						Compaction: tsdb.BlockMetaCompaction{
+							Sources: []ulid.ULID{ULID(2)},
+						},
+					},
+				},
+				ULID(3): {
+					BlockMeta: tsdb.BlockMeta{
+						ULID: ULID(3),
+						Compaction: tsdb.BlockMetaCompaction{
+							Sources: []ulid.ULID{ULID(3)},
+						},
+					},
+				},
+			},
+			expected: func(input map[ulid.ULID]*metadata.Meta) map[ulid.ULID]*metadata.Meta {
+				return map[ulid.ULID]*metadata.Meta{
+					ULID(1): input[ULID(1)],
+					ULID(2): input[ULID(2)],
+					ULID(3): input[ULID(3)],
+				}
+			},
+		},
+		{
+			name: "compacted meta in bucket",
+			input: map[ulid.ULID]*metadata.Meta{
+				ULID(4): {
+					BlockMeta: tsdb.BlockMeta{
+						ULID: ULID(4),
+						Compaction: tsdb.BlockMetaCompaction{
+							Sources: []ulid.ULID{ULID(1), ULID(2), ULID(3), ULID(4)},
+						},
+					},
+				},
+			},
+			expected: func(input map[ulid.ULID]*metadata.Meta) map[ulid.ULID]*metadata.Meta {
+				return map[ulid.ULID]*metadata.Meta{
+					ULID(4): input[ULID(4)],
+				}
+			},
+		},
+		{
+			name: "two blocks with same sources",
+			input: map[ulid.ULID]*metadata.Meta{
+				ULID(3): {
+					BlockMeta: tsdb.BlockMeta{
+						ULID: ULID(3),
+						Compaction: tsdb.BlockMetaCompaction{
+							Sources: []ulid.ULID{ULID(1), ULID(2)},
+						},
+					},
+				},
+				ULID(4): {
+					BlockMeta: tsdb.BlockMeta{
+						ULID: ULID(4),
+						Compaction: tsdb.BlockMetaCompaction{
+							Sources: []ulid.ULID{ULID(1), ULID(2)},
+						},
+					},
+				},
+			},
+			expected: func(input map[ulid.ULID]*metadata.Meta) map[ulid.ULID]*metadata.Meta {
+				return map[ulid.ULID]*metadata.Meta{
+					ULID(3): input[ULID(3)],
+				}
+			},
+		},
+		{
+			name: "two blocks with overlapping sources",
+			input: map[ulid.ULID]*metadata.Meta{
+				ULID(3): {
+					BlockMeta: tsdb.BlockMeta{
+						ULID: ULID(3),
+						Compaction: tsdb.BlockMetaCompaction{
+							Sources: []ulid.ULID{ULID(3)},
+						},
+					},
+				},
+				ULID(4): {
+					BlockMeta: tsdb.BlockMeta{
+						ULID: ULID(4),
+						Compaction: tsdb.BlockMetaCompaction{
+							Sources: []ulid.ULID{ULID(1), ULID(2)},
+						},
+					},
+				},
+				ULID(5): {
+					BlockMeta: tsdb.BlockMeta{
+						ULID: ULID(5),
+						Compaction: tsdb.BlockMetaCompaction{
+							Sources: []ulid.ULID{ULID(1), ULID(2), ULID(3)},
+						},
+					},
+				},
+			},
+			expected: func(input map[ulid.ULID]*metadata.Meta) map[ulid.ULID]*metadata.Meta {
+				return map[ulid.ULID]*metadata.Meta{
+					ULID(5): input[ULID(5)],
+				}
+			},
+		},
+		{
+			name: "3 non compacted metas and compacted meta of level 2 in bucket",
```

Comment:
again, meta is never technically compacted -> we need to consistent and clear in naming here. Also compaction does not matter here: Just what constains same sources.

For example this case could be named: `Block 4 contains 1, 2, 3`

Also I believe :
```
ULID: ULID(4),
						Compaction: tsdb.BlockMetaCompaction{
							Sources: []ulid.ULID{ULID(1), ULID(2), ULID(3), ULID(4)},
						},
```
.. is wrong. This will never happen as 4 is compacted block, so it will be never part of sources I believe (That's why I linked the Prometheus code which does it, to double check) (: